Maturing the extensions pattern (#508)
gruetter/innerSourceTest: patterns/1-initial/extensions-for-sustainable-growth.md

## Title

> Extensions for Sustainable Growth

Reviewer: I like the new title!

We can also add the old title as an "Alias" at the bottom of the pattern.
That way the pattern can also be found via those keywords.

## Patlet

An InnerSource project is receiving too many contributions, making maintenance difficult. By offering an extension mechanism outside of the core project, the maintainers enable scaling of project capabilities with minimal cost and maintenance overhead.

## Problem

> As the number of contributions to a mature InnerSource repository rapidly increases, it adds more burden on code reviews and maintenance. This results in a large code review backlog or premature rejection of new feature contributions. How do we allow for faster release of new features, encouraging innovation and experimentation; while also keeping the repository well maintained?

Reviewer: ```suggestion
As the number of contributions to a mature InnerSource repository rapidly increases, it adds more burden on code reviews and maintenance. This results in a large code review backlog or premature rejection of new feature contributions.

How can the host team allow for faster release of new features, encouraging innovation and experimentation; while also keeping the repository well maintained?
```

Replaced "we" with "host team", to make it more clear who is pushing for this improvement.

## Story

> There is a strategic project that aims to collect the best innovations within a domain space to one common stack, allowing reuse of a common infrastructure and providing a standard user experience. Through InnerSource, various teams in the organization that work within the domain space get an opportunity to collaborate and contribute their innovations to the common codebase. However, a large number of contributions in parallel from several developers is making maintenance of the codebase difficult. This is adding a huge burden on the project maintainers who assume ownership over the code quality standards and enable the community through various forms of communication.

Reviewer: ```suggestion
There is a strategic project that aims to collect the best innovations within a domain space to one common stack, allowing reuse of a common infrastructure and providing a standard user experience. Through InnerSource, various teams in the organization that work within the domain space get an opportunity to collaborate and contribute their innovations to the common codebase.

However, a large number of contributions in parallel from several developers is making maintenance of the codebase difficult. This is adding a huge burden on the project maintainers who assume ownership over the code quality standards and enable the community through various forms of communication.
```

Seems like a natural place to split this rather long paragraph.

> Project maintainers are burnt out due to:

Reviewer: Do you literally mean "burnout" here, or could we use a different phrase?

Just wondering if burnout might be too strong? 
However if that is how it was for the group in your story, by all means, keep it as is.

Author: I think burnout would be an appropriate term here that would be caused because of reasons mentioned. I could also remove this paragraph , since the matter is described above `However, a large number of contributions in parallel from several developers is making maintenance of the codebase difficult. This is adding a huge burden on the project maintainers who assume ownership over the code quality standards and enable the community through various forms of communication.`

and essentially we would like extensions pattern to be applied before maintainers face burnout , when maintenance starts to become difficult for a project

Reviewer: Feel free to leave the burnout term in there, if this is the extent that you observed in your story.

We could also change it to this. Might make it a bit less strong?
> Project maintainers are at risk of burnout due to:

Author: Good idea, updated

> - Everlasting backlog of pull requests that need to be reviewed.

Reviewer: ```suggestion
- Everlasting backlog of pull requests from contributors that need to be reviewed.
```

- Job dissatisfaction: Majority of maintainers' time spent in community support leaves no room for innovation.
- Perceived lack of accomplishment: Not all contributed features have adequate user demand and result in consequent adoption.
- Time consuming releases: More features in the codebase results in long running tests.
- Increase in maintenance activities: More bugs raised as new capabilities are added.

> A lot of time and investment is spent on maturing every new feature contribution, before potential users even get an opportunity to explore the features for their use cases.

Reviewer: ```suggestion
A lot of time is spent on maturing every new feature contribution, before potential users even get an opportunity to explore the feature for their use cases. If it turns out that new feature isn't fulfilling the use case, then all that time spent on achieving the desired code quality standards are waste.
```

I elaborated on your sentence here to make it more explicit what the waste is.

## Context

- A strategic InnerSource codebase is scaling rapidly with new feature contributions from several employees.
- The ratio of reviewers to contributions results in a growing backlog of pull requests. This is slowing down release of new features to the community.

> - Quality of the codebase is no longer maintained and user experience is adversely impacted.

Reviewer: ```suggestion
- Quality of the codebase is degrading and user experience is adversely impacted.
```

- Maintainers of the codebase are burdened and cannot keep up with the influx of contributions and increased community support.

> - There are more features being contributed than gaining adoption by users. Not all ideas are serving internal use cases, but addition of features is adding to the maintenance overhead.

Reviewer: ```suggestion
- Some of the contributed features are not gaining adoption by users, and might even turn fully dormant. However even though they are unused, these features are still adding to the maintenance overhead.
```

Would this be a simpler way to say the same thing?

## Solutions

Allowing [extensions/plugins](https://en.wikipedia.org/wiki/Extensibility) to high-scale InnerSource codebases can relieve the maintenance burden on repository maintainers and allow faster release of new features for adopting products to explore. This shifts maintenance of capabilities to extension owners and allows the primary repository to support capabilities that have been adopted more widely and are more strategic.  

> Extensions provide a filter for new capabilities that may eventually move into the core of the project. Extensions also act as an incubation and community hardening environment, allowing for much of that hardening to happen organically rather than in a costly review process.

Reviewer: What do you mean by "filter"?

Author: hmm so not all new capabilities will move into the primary repository and project. Only ones that meet users needs. So that way extensions are filtering out what new capabilities make it into core of project.

Reviewer: Got it. Let's leave it as is.

   - Create criteria for when an extension can be ported to the primary repository, such as adoption of the extension by internal products and demand for the feature.
   - Porting process of the extension to the primary repository will follow more stringent code review guidelines as set by library maintainers.

> ![Project extensions](../../assets/img/extensions-for-sustainable-growth/extensions-for-sustainable-growth.png)

Reviewer: In the published pattern the text "Project extensions" will be displayed as a caption below the image.
So please put here whatever you want to be shown for that purpose.

> 
> Following the principles outlined above ensures that:

Reviewer: ```suggestion

Following these principles ensures that:
```

simplifying

> - The addition of the extensions pattern with easy to use templates enables developers to add new features to a project's ecosystem without requiring them to write large amounts of [boilerplate](https://en.wikipedia.org/wiki/Boilerplate_code) code.

Reviewer: ```suggestion
- Developers can add new features to a project's ecosystem without requiring them to write large amounts of [boilerplate](https://en.wikipedia.org/wiki/Boilerplate_code) code.
```

I would not talk about the "pattern" at this point, as that is not how they users would think about it.
They might think "I added an extension using a template" but probably not "I used the extension pattern" :)

> - Extensions are discoverable in a repeatable manner to all users of the primary project; just because code doesn't live in the main repository yet does not mean it is not valuable.

Reviewer: ```suggestion
- Extensions are discoverable in a repeatable manner by all users of the primary project; just because code doesn't live in the main repository yet does not mean it is not valuable.
```

- The maintainer burden is reduced until an extension has demonstrated that it fills an important gap in the primary project.
- The core project's common code (e.g. base classes and utility functions) can be a starting point for new development that extends project's domain. This avoids the need to port innovative work after-the-fact, thus reducing the overall burden of developing novel features for the project.
- Developers are more likely to contribute and stay involved in maintenance and building communities for their codebase, which is also good for the health of the overall project ecosystem.

## Resulting Context

> - The project is able to scale with the addition of new features, without adding maintenance overhead on the primary project repository.

Reviewer: Is there really zero maintenance work created when using the extensions mechanism?
Or is it rather that the maintenance that is required for this to work seems more fruitful and scales better?

Author: since extensions lie outside project, they would not add a maintenance overhead to primary project repository . They will be independent repositories with the new features

> - Faster release of new features and experimental features for the community to explore, encouraging innovation and experimentation.

Reviewer: Do you mean "new features" via the core and "experimental features" via the extensions?

Author: I will remove experimental features here, since all new features are meant for exploration/ experimentation

## Known Instances

> **IBM Corporation** has adopted this solution to scale [InnerSource AI libraries](https://youtu.be/Lz-tIc2cyRM). Using extensions, developers are able to extend AI libraries with more algorithms and share their innovations with the community. The core libraries only contain strategic algorithms that have been adopted and validated, keeping them easier to maintain as we scale contributions.

Reviewer: ```suggestion
* **IBM Corporation** has adopted this solution to scale [InnerSource AI libraries](https://youtu.be/Lz-tIc2cyRM). Using extensions, developers are able to extend AI libraries with more algorithms and share their innovations with the company-internal community. The core libraries only contain strategic algorithms that have been adopted and validated, keeping them easier to maintain as we scale contributions.

## Alias

Extensions to Manage Contributions at Scale
```

Let's make this a 1-bullet list, to highlight that we want to collect a list of known instances here.

Also added the Alias that I mentioned before.

> ## Status

Reviewer: You can already change the status to **Structured**.

Once all comments are resolved, you can then `git mv` the file to the `2-structured` folder as well.

Author: @spier just few unresolved comments left, let me know what you think and if answers are clear. I will then merge the PR and move to structured. Thank you
